src/audio/mixer.c

```diff
@@ -250,41 +261,44 @@ static int mixer_copy(struct comp_dev *dev)
 	if (num_mix_sources == 0)
 		return 0;
 
-	/* make sure no sources have underruns */
-	for (i = 0; i < num_mix_sources; i++) {
+	/* check for overrun */
+	if (sink->free == 0) {
+		trace_mixer_error("mixer_copy() error: sink component buffer "
+				  "has not enough free bytes for copy");
+		comp_overrun(dev, sink, 0, 0);
+		return -EIO;
+	}
 
-		/* make sure source component buffer has enough data available
-		 * and that the sink component buffer has enough free bytes
-		 * for copy. Also check for XRUNs
-		 */
-		res = comp_buffer_can_copy_bytes(sources[i], sink, md->period_bytes);
-		if (res < 0) {
-			trace_mixer_error("mixer_copy() error: "
-					  "source component buffer "
-					  "has not enough data available");
-			comp_underrun(dev, sources[i], sources[i]->avail,
-				      md->period_bytes);
-		} else if (res > 0) {
-			trace_mixer_error("mixer_copy() error: "
-					  "sink component buffer has not "
-					  "enough free bytes for copy");
-			comp_overrun(dev, sources[i], sink->free,
-				     md->period_bytes);
+	/* check for underruns */
+	for (i = 0; i < num_mix_sources; i++) {
+		if (sources[i]->avail == 0) {
+			trace_mixer_error("mixer_copy() error: source %u "
+					  "component buffer has not enough "
+					  "data available", i);
+			comp_underrun(dev, sources[i], 0, 0);
+			return -EIO;
 		}
+
+		frames = MIN(frames, comp_avail_frames(sources[i], sink));
 	}
 
+	source_bytes = frames * comp_frame_bytes(sources[0]->source);
+	sink_bytes = frames * comp_frame_bytes(sink->sink);
+
+	tracev_mixer("mixer_copy(), source_bytes = 0x%x, sink_bytes = 0x%x",
+		     source_bytes, sink_bytes);
+
 	/* mix streams */
-	md->mix_func(dev, sink, sources, i, dev->frames);
+	md->mix_func(dev, sink, sources, i, frames);
 
-	/* update source buffer pointers for overflow */
+	/* update source buffer pointers */
 	for (i = --num_mix_sources; i >= 0; i--)
-		comp_update_buffer_consume(sources[i], md->period_bytes);
+		comp_update_buffer_consume(sources[i], source_bytes);
```

Comment:
@tlauda what is the assumption here in using only the source_bytes calculated for source[0] and consuming the same number of bytes from all the sources?

Reply:
@ranj063 Mixer expects the same format from all the sources so we can base our calculations just on one of the sources. We are calculating the min bytes, so the consumption value would be valid for all of them.

Reply:
@tlauda then can you please add a comment here stating this.
Also, you need to add a check to make sure this is indeed the case.